### PR TITLE
Leverage az provider list

### PR DESCRIPTION
**Pull request by danielrbradley**
Fixes https://github.com/pulumi/pulumi-azure-native/issues/2661

1. Add inactive default versions report - which versions referenced from our default version are not currently marked as "live" by `az provider list`?
2. When picking a new tracking version, select from the live versions only, if any are live. Fall back to all versions if `az provider list` doesn't include any.
3. When checking for additional resources not in the default version, don't add new non-live versions.
4. Move most unit testing into gen_calculateVersionMetadata_test from defaultVersion_test as it's as close to e2e as possible and requires less mocking.

### CODE REVIEW

**Review by thomas11 — COMMENTED** [review]
Pretty tricky to review but overall LGTM. Thanks for the nice test coverage!

> Fall back to all versions if as provider list doesn't include any.

Is this in case `az provider list` is wrong? If it was 100% correct, we could disregard that resource altogether, no?

My main concern is that I don't think we should merge the schema and SDKs with these changes in v2. Even if the Azure API versions are not actually live, it's still a breaking code change. And there's a small chance `provider list` is wrong.

**Comment by danielrbradley** [comment]
> My main concern is that I don't think we should merge the schema and SDKs with these changes in v2. Even if the Azure API versions are not actually live, it's still a breaking code change. And there's a small chance `provider list` is wrong.

Agreed. I was also concerned about removing all these. Perhaps we should feature switch the additions removals off until V3.

**Comment by danielrbradley** [comment]
@thomas11 Ok, latest set of changes from the original design are up.

Instead of feature switching the removal of the non-live additions, I've just changed the logic slightly to not remove additions which are already part of the spec - even if they're not live. This now results in zero schema changes.

This maintains the current set of default resources but will naturally kick in for v3 when we generate a new default version from scratch.

File: Makefile [hunk 1]
```diff
@@ -112,8 +112,8 @@ update_submodules:
 		echo "Updating submodule $$submodule" ; \
 		(cd $$submodule && git checkout main && git pull origin main); \
 	done
-	rm ./azure-provider-versions/provider_list.json
-	az provider list | jq 'map({ namespace: .namespace, resourceTypes: .resourceTypes | map({ resourceType: .resourceType, apiVersions: .apiVersions }) | sort_by(.resourceType) }) | sort_by(.namespace)' > ./azure-provider-versions/provider_list.json
+	rm versions/az-provider-list.json
+	az provider list --query 'sort_by([*].{ namespace: namespace, resourceTypes: sort_by(resourceTypes[*].{ resourceType: resourceType, defaultApiVersion: defaultApiVersion, apiVersions: apiVersions, locations: locations }, &resourceType) }, &namespace)' > versions/az-provider-list.json
```
Comment (thomas11) [inline Makefile:116]: This line could use a comment

File: provider/pkg/providerlist/providerlist.go [hunk 2]
```diff
@@ -24,8 +24,24 @@ type ApiVersion = string
 // ProviderPathVersions is a map of lowered provider names to Api Versions e.g. `analysisservices -> "locations/checknameavailability" -> [2019-03-01-preview, 2018-05-05]`
 type ProviderPathVersions = map[LoweredProviderName]map[ResourcePath]ApiVersions
 
+type ResourceType struct {
```
Comment (thomas11) [inline provider/pkg/providerlist/providerlist.go:27]: Short doc comments for public types? I'm not sure what a "ResourceType" is without reading more code.

**Review by thomas11 — APPROVED** [review]
LGTM with some docs wishlist items

File: provider/pkg/providerlist/providerlist.go (hunk 2 above)
Comment (danielrbradley) [reply provider/pkg/providerlist/providerlist.go:29]: Good thoughts. I've added via a rebase 👍

File: Makefile (hunk 1 above)
Comment (danielrbradley) [reply Makefile:117]: Done 👍

**Pull request merged**